src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
+
+        public DistributedLockService(IMachineNameProvider machineNameProvider, ILifetimeScope lifetimeScope, IClock clock) {
+            _machineNameProvider = machineNameProvider;
+            _lifetimeScope = lifetimeScope;
+            _clock = clock;
+            Logger = NullLogger.Instance;
+        }
+
+        public ILogger Logger { get; set; }
+
+        public bool TryAcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout, out IDistributedLock @lock) {
+            lock (_semaphore) {
+                @lock = default(IDistributedLock);
+
+                try {
+                    var waitedTime = TimeSpan.Zero;
+                    var waitTime = TimeSpan.FromMilliseconds(timeout.TotalMilliseconds / 10);
+                    bool acquired;
+
+                    while (!(acquired = TryAcquireLockRecord(name, maxLifetime, out @lock)) && waitedTime < timeout) {
+                        Task.Delay(timeout).ContinueWith(t => {
+                            waitedTime += waitTime;
+                        }).Wait();
+                    }
+
+                    if (acquired) {
+                        Logger.Debug("Successfully acquired a lock named {0}.", name);
+                        return true;
+                    }
+                }
+                catch (Exception ex) {
+                    Logger.Error(ex, "Error while trying to acquire a lock named {0}.", name);
+                    throw;
+                }
+
+                Logger.Debug("Could not acquire a lock named {0}.", name);
+                return false;
+            }
+        }
+
+        public IDistributedLock AcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout) {
+            IDistributedLock lockResult;
+            return TryAcquireLock(name, maxLifetime, timeout, out lockResult) ? lockResult : null;
+        }
+
+        public void DisposeLock(IDistributedLock @lock) {
+            var childLifetimeScope = CreateChildLifetimeScope(@lock.Name);
+
+            try {
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                var record = repository.Get(@lock.Id);
+
+                if (record != null) {
+                    if (record.ReferenceCount > 0)
+                        record.ReferenceCount--;
+                }
```

Comment:
I think this should throw if you're trying to release the same lock too many times, or trying to release a lock that doesn't exist. Otherwise might be obscuring bugs in calling code.